### ISSUE TITLE
Bahá'í date functions

### ISSUE THREAD

**Issue by dragfyre** [bahaidev/bahaibot#10]
Add the capability to convert to/from Bahá'í dates. Should include the following:
- `!date` command that auto-detects a date in Badi' or Gregorian format and converts to the opposite format
- Toggle to display the Bahá'í date at the top of the server's channel list (off by default except on Bahá'í.FYI / Baha'i CoLab), similar to the following screenshot:

![Screen Shot 2021-06-06 at 3 44 46 PM](https://user-images.githubusercontent.com/4483894/120918330-26bea600-c6de-11eb-9c7c-f1aef221211c.png)

**Comment by brettz9** [comment]
I've added support for `/date`/`!date` to convert Gregorian to Badí'. For Badí' to Gregorian, we need to add parsing support to the likes of `bahai-date-api`.

**Comment by brettz9** [comment]
I've added parsing support, so now conversions are in two directions.

**Comment by dragfyre** [comment]
What do you think of the second item - getting the date to show up in a locked channel at the top of the channel list? This will probably be more complicated than the first item, so I think it's OK if we split it off to a new issue.

**Comment by brettz9** [comment]
Before I can test what I've done to try to do that, we need to have the View and Manage Channels permissions for the bot on the wikis server, on Lab, and on FYI.

**Comment by dragfyre** [comment]
A few suggestions for the date channel, as noted on Discord:
- It would be best to **use a voice channel** and not a text channel, because voice channels can use spaces in their title
- Title: leave off "bahai date" at the start, just show the date
  - Format can be <day_number> <arabic_month_name>/<english_month_name> <bahai_year> BE
- Channel should have **"connect" and "speak" perms off** so that people can't use the channel to speak
- On bot launch, bot should check for an existing voice channel with the pattern "<digits><single_space>BE" at the end of the title
  - If such a channel exists, use it as the date channel; if not, create one and use it

**Comment by dragfyre** [comment]
Starting to look nice:

<img width="285" height="87" alt="Image" src="https://github.com/user-attachments/assets/020c2404-cc57-41e3-bbcd-113d04b24dd3" />

**Comment by dragfyre** [comment]
For what it's worth, I checked that `/date` can handle various date formats, and it can:

<img width="359" height="582" alt="Image" src="https://github.com/user-attachments/assets/bf68a147-9f6d-4a14-8b7c-2f54c70be42d" />

**Comment by dragfyre** [comment]
Seems like the Badi date formats aren't working though. (Same null result when leaving out "BE" at the end.) Did we forget something? 🤔

<img width="328" height="146" alt="Image" src="https://github.com/user-attachments/assets/03734bd1-17bd-4e85-89a7-cf3f2c66a893" />

**Comment by brettz9** [comment]
There is no proper parser for Badí' dates, so I just have a regular expression which currently only handles dates in these formats:

```
15 Qawl (Speech), 100 (Asia/Jerusalem)
15 Qawl, 100 (Asia/Jerusalem)
```

I've now added support for
1. dropping the parenthetical timezone portion (defaulting to "Asia/Jerusalem").
2. an optional " BE"

**Comment by dragfyre** [comment]
> dropping the parenthetical timezone portion (defaulting to "Asia/Jerusalem").

Agreed, this is a good way of handling it - I'd say 99.995% of people would not bother to add a timezone reference at the end. (Basically just Brett hehe) Also, I would suggest making the comma before the year optional as well.

For the mixed Arabic/English format ("15 Qawl (Speech), 100"), what would you think of instead matching just the English? That way we would have two different formats to match:

- English: 15 Speech, 100
- Arabic: 15 Qawl, 100

**Comment by brettz9** [comment]
> > dropping the parenthetical timezone portion (defaulting to "Asia/Jerusalem").
> 
> Agreed, this is a good way of handling it - I'd say 99.995% of people would not bother to add a timezone reference at the end. (Basically just Brett hehe) 

:-)

> Also, I would suggest making the comma before the year optional as well.

Yeah, just did that.
 
> For the mixed Arabic/English format ("15 Qawl (Speech), 100"), what would you think of instead matching just the English? That way we would have two different formats to match:
> 
> * English: 15 Speech, 100
> * Arabic: 15 Qawl, 100

The implementation should already do this.

**Comment by dragfyre** [comment]
> The implementation should already do this.

OK - I'll check when the bot is back up. (Seems to be a lot of crashes lately as mentioned here: https://github.com/bahaidev/bahaibot/issues/17#issuecomment-4456292449)

**Comment by brettz9** [comment]
The bot is back up.
